src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh

```diff
@@ -0,0 +1,410 @@
+    zones_.push_back(nucl_zones);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return the number of nuclear zones.
+ */
+size_type NuclearZoneBuilder::num_nuclear_zones(AtomicMassNumber a)
+{
+    return (a.get() < 5) ? 1 : (a.get() < 100) ? 3 : 6;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate components of nuclear zone data: the nuclear zone radius, volume,
+ * density, Fermi momentum and potential function as in G4NucleiModel and as
+ * documented in section 24.2.3 of the Geant4 Physics Reference (release 11.2).
+ */
+CELER_FUNCTION
+void NuclearZoneBuilder::calc_zone_component(IsotopeView const& target,
+                                             ComponentVec& components)
+{
+    AtomicNumber a = target.atomic_mass_number();
+
+    // Calculate nuclear radius
+    real_type nuclear_radius = this->calc_nuclear_radius(a);
+
+    real_type skin_ratio = nuclear_radius / skin_depth_;
+    real_type skin_decay = std::exp(-skin_ratio);
+
+    // Temporary data for the zone-by-zone density function
+    size_type num_of_zones = components.size();
+    std::vector<real_type> ur(num_of_zones + 1);
+    std::vector<real_type> integral(num_of_zones);
+    real_type total_integral{0};
+
+    // Fill the nuclear radius by each zone
+    auto amass = a.get();
+    ur[0] = (amass < 12) ? 0 : -skin_ratio;
+
+    if (amass < 5)
+    {
+        // Light ions treated as simple balls
+        components[0].radius = nuclear_radius;
+        integral[0] = 1;
+        total_integral = integral[0];
+    }
+    else if (amass < 100)
+    {
+        constexpr Real3 alpha = {0.7, 0.3, 0.01};
+        if (amass < 12)
+        {
+            // Small nuclei have a three-zone Gaussian potential
+            real_type gauss_radius
+                = std::sqrt(ipow<2>(nuclear_radius)
+                                * (1 - 1 / static_cast<real_type>(amass))
+                            + real_type{6.4});
+            for (auto i : range(num_of_zones))
+            {
+                real_type y = std::sqrt(-std::log(alpha[i]));
+                components[i].radius = gauss_radius * y;
+                ur[i + 1] = y;
+                integral[i]
+                    = this->integrate_gaussian(ur[i], ur[i + 1], gauss_radius);
+                total_integral += integral[i];
+            }
+        }
+        else
+        {
+            // Intermediate nuclei have a three-zone Woods-Saxon potential
+            for (auto i : range(num_of_zones))
+            {
+                real_type y = std::log((1 + skin_decay) / alpha[i] - 1);
+                components[i].radius = nuclear_radius + skin_depth_ * y;
+                ur[i + 1] = y;
+                integral[i] = this->integrate_woods_saxon(
+                    ur[i], ur[i + 1], nuclear_radius);
+                total_integral += integral[i];
+            }
+        }
+    }
```

Comment:
This would be easier to read with 
```c++
    else if (amass < 12)
    {
        constexpr Real3 sqrt_minus_log_alpha = {/* precompute these*/};
        /* * */
    }
    else if (amass < 100)
    {
        constexpr Real3 alpha = {0.7, 0.3, 0.01};
        /* * */
    }
```

Reply:
Okay, I did not through about this pre-computation during refactoring Geant4 codes.  Good suggestion. Changed accordingly.